File: src/lib/BaseSuite.ts
/* eslint-disable @typescript-eslint/no-unused-vars */

// Do not add logging to this file as it is used by the pure runtime.

import { ITTestResourceConfiguration, ITestArtifactory } from ".";
import { Ibdd_in_any, Ibdd_out_any } from "../CoreTypes";
import { IGivens } from "./abstractBase";
import { beforeAllProxy, afterAllProxy } from "./pmProxy";
import { IPM } from "./types";

export abstract class BaseSuite<I extends Ibdd_in_any, O extends Ibdd_out_any> {
  name: string;
  givens: IGivens<I>;
  store: I["istore"];
  testResourceConfiguration: ITTestResourceConfiguration;
  index: number;
  failed: boolean;
  fails: number;

  artifacts: string[] = [];

  addArtifact(path: string) {
    const normalizedPath = path.replace(/\\/g, "/"); // Normalize path separators
    this.artifacts.push(normalizedPath);
  }

  constructor(name: string, index: number, givens: IGivens<I> = {}) {
    const suiteName = name || "testSuite"; // Ensure name is never undefined
    if (!suiteName) {
      throw new Error("BaseSuite requires a non-empty name");
    }

    this.name = suiteName;
    this.index = index;
    this.givens = givens;
    this.fails = 0;
  }

  public features() {
    try {
      const features = Object.keys(this.givens)
        .map((k) => this.givens[k].features)
        .flat()
        .filter((value, index, array) => {
          return array.indexOf(value) === index;
        });
      return features || [];
    } catch (e) {
      console.error("[ERROR] Failed to extract features:", JSON.stringify(e));
      return [];
    }
  }

  public toObj() {
    const givens = Object.keys(this.givens).map((k) => this.givens[k].toObj());

    return {
      name: this.name,
      givens,
      fails: this.fails,
      failed: this.failed,
      features: this.features(),
    };
  }

  setup(
    s: I["iinput"],
    artifactory: ITestArtifactory,
    tr: ITTestResourceConfiguration,
    pm: IPM
  ): Promise<I["isubject"]> {
    return new Promise((res) => res(s as unknown as I["isubject"]));
  }

  assertThat(t: Awaited<I["then"]> | undefined): boolean {
    return !!t;
  }

  afterAll(store: I["istore"], artifactory: ITestArtifactory, pm: IPM) {
    return store;
  }

  async run(
    input: I["iinput"],
    testResourceConfiguration: ITTestResourceConfiguration,
    artifactory: (fPath: string, value: unknown) => void,
    tLog: (...string) => void,
    pm: IPM
  ): Promise<BaseSuite<I, O>> {
    this.testResourceConfiguration = testResourceConfiguration;
    // tLog("test resources: ", JSON.stringify(testResourceConfiguration));

    const suiteArtifactory = (fPath: string, value: unknown) =>
      artifactory(`suite-${this.index}-${this.name}/${fPath}`, value);

    // console.log("\nSuite:", this.index, this.name);
    // tLog("\nSuite:", this.index, this.name);
    const sNdx = this.index;
    // Ensure addArtifact is properly bound to 'this'
    const addArtifact = this.addArtifact.bind(this);
    const proxiedPm = beforeAllProxy(pm, sNdx.toString(), addArtifact);
    const subject = await this.setup(
      input,
      suiteArtifactory,
      testResourceConfiguration,
      proxiedPm
    );

    for (const [gKey, g] of Object.entries(this.givens)) {
      const giver = this.givens[gKey];
      this.store = await giver
        .give(
          subject,
          gKey,
          testResourceConfiguration,
          this.assertThat,
          suiteArtifactory,
          tLog,
          pm,
          sNdx
        )
        .catch((e) => {
          this.failed = true;
          this.fails = this.fails + 1;
          throw e;
        });
    }

    try {
      // Ensure addArtifact is properly bound to 'this'
      const addArtifact = this.addArtifact.bind(this);
      const afterAllPm = afterAllProxy(pm, sNdx.toString(), addArtifact);
      this.afterAll(this.store, artifactory, afterAllPm);
    } catch (e) {
      console.error(JSON.stringify(e));
      // this.fails.push(this);
      // return this;
    }

    return this;
  }
}

File: src/lib/rusto/src/base_suite.rs
use crate::types::{IbddInAny, IbddOutAny, ITTestResourceConfiguration};
use std::collections::HashMap;

pub struct BaseSuite<I: IbddInAny, O: IbddOutAny> {
    pub name: String,
    pub givens: HashMap<String, Box<dyn std::any::Any>>,
    pub store: Option<I::Istore>,
    pub test_resource_configuration: Option<ITTestResourceConfiguration>,
    pub index: usize,
    pub failed: bool,
    pub fails: i32,
    pub artifacts: Vec<String>,
}

impl<I: IbddInAny, O: IbddOutAny> BaseSuite<I, O> {
    pub fn new(name: String, givens: HashMap<String, Box<dyn std::any::Any>>) -> Self {
        Self {
            name,
            givens,
            store: None,
            test_resource_configuration: None,
            index: 0,
            failed: false,
            fails: 0,
            artifacts: Vec::new(),
        }
    }
    
    pub fn add_artifact(&mut self, path: String) {
        let normalized_path = path.replace('\\', "/");
        self.artifacts.push(normalized_path);
    }
    
    pub fn features(&self) -> Vec<String> {
        let mut features = Vec::new();
        let mut seen = std::collections::HashSet::new();
        
        for given in self.givens.values() {
            // In a real implementation, we would extract features from the given
            // For now, return empty vector
        }
        
        features
    }
    
    pub fn to_obj(&self) -> HashMap<String, Box<dyn std::any::Any>> {
        let mut obj = HashMap::new();
        obj.insert("name".to_string(), Box::new(self.name.clone()) as Box<dyn std::any::Any>);
        obj.insert("fails".to_string(), Box::new(self.fails) as Box<dyn std::any::Any>);
        obj.insert("failed".to_string(), Box::new(self.failed) as Box<dyn std::any::Any>);
        obj.insert("features".to_string(), Box::new(self.features()) as Box<dyn std::any::Any>);
        obj.insert("artifacts".to_string(), Box::new(self.artifacts.clone()) as Box<dyn std::any::Any>);
        obj
    }
    
    pub async fn run(
        &mut self,
        input: I::Iinput,
        test_resource_configuration: ITTestResourceConfiguration,
    ) -> &Self {
        self.test_resource_configuration = Some(test_resource_configuration);
        self.fails = 0;
        self.failed = false;
        
        // In a real implementation, this would run all givens
        // For now, just return self
        
        self
    }
}
